composer/datasets/in_context_learning_evaluation.py

```diff
@@ -109,12 +198,193 @@ def update_metric(self, metric, batch, output_logits, labels):
+                query = f'{example_delimiter}{query}'
+
+            choices = [f'{continuation_delimiter}{choice}' for choice in choices]
+            encoded_example['query'] = self.tokenizer(query)
+            encoded_example['choices'] = [self.tokenizer(choice) for choice in choices]
+            encoded_example['preamble'] = self.tokenizer(
+                preamble)  # if the preamble is empty then these will be 0-length lists
+            encoded_example['gold_idx'] = gold_idx
+
+            examples.append(encoded_example)
+
+        return examples
+
+    def __getitem__(self, index):
+        return self.encoded_dataset[index]
+
+    def __len__(self):
+        return len(self.encoded_dataset)
+
+    def collate_fn(self, data):
+        inputs = []
+        continuation_indices = []
+        gold_idxs = []
+        choice_groupings = []
+        for data_pair in data:
+
+            choice_start_idx = len(continuation_indices)
+            preamble, context, choices, gold_idx = (data_pair['preamble'], data_pair['query'], data_pair['choices'],
+                                                    data_pair['gold_idx'])
+
+            for choice in choices:
+                context_enc = preamble['input_ids'] + context['input_ids']
+                continuation_enc = choice['input_ids']
+                inp, continuation_span = _make_padded_input(context_enc, continuation_enc, self.max_seq_len,
+                                                            self.eos_tok_id)
+
+                inputs.append(inp)
+                continuation_indices.append(continuation_span)
+
+            gold_idxs.append(gold_idx)
+            choice_end_idx = len(continuation_indices)
+            choice_groupings.append((choice_start_idx, choice_end_idx))
+
+        # We run each distinct query + answer choice through the model separately and determine which
+        # answer has the lowest per-token-perplexity.
+        #
+        # If each question has N possible choices, all N must be grouped together as distinct elements of the batch
+        # since the batch may consist of multiple questions, the choice_groupings indicates
+        # which contiguous sequences of elements in the batch correspond to which question
+        # gold_indices indicates which of the [0, N-1] choices is the correct one for each question.
+        batch = {
+            'input_ids': torch.stack(inputs),
+            'continuation_indices': continuation_indices,
+            'mode': 'icl_task',
+            'labels': torch.stack(inputs),
+            'gold_indices': gold_idxs,
+            'choice_groupings': choice_groupings
+        }
+        batch['attention_mask'] = ~(batch['input_ids'] == self.eos_tok_id)
+        return batch
+
+    def get_num_samples_in_batch(self, batch) -> int:
+        return batch['input_ids'].shape[0]
+
+    def update_metric(self, metric, batch, output_logits, labels):
+        metric.update(batch, output_logits, labels)
```

Comment:
Are these functions necessary? I thought Composer could handle `get_num_samples_in_batch` HF style batches automatically.

The `update_metric` doesn't seem to be used.

Reply:
I'm guessing that the default ends up raising here https://github.com/mosaicml/composer/blob/c99230435cd23728e9e5d8cbdc62bc92cbbaa64d/composer/core/data_spec.py#L297 or maybe even crashing here (https://github.com/mosaicml/composer/blob/c99230435cd23728e9e5d8cbdc62bc92cbbaa64d/composer/core/data_spec.py#L292) because of the non tensor items in the dictionary. Possible the default should get slightly smarter (i.e. look for `input_ids`, or skip non tensor items in the batch). Made a JIRA to look into this https://mosaicml.atlassian.net/browse/CO-1660.

Reply:
`update_metric` is what composer calls to update metrics. It isn't implemented on the base class (maybe it should be?), so any subclass of `ComposerModel` that wants to do eval needs to implement it.

Reply:
yep you're right about the update metric function being superfluous and needing to delete it. Also Daniel is right about why the default doesn't work!